MD Tag parsing errors during creation of DecadentRead in BQSR

Reported by @beaunorgeot via email. Exception, with relevant stack trace is:

```
java.lang.IllegalArgumentException: Error "(D,Some(3)) (of class scala.Tuple2)" while constructing DecadentRead from Read({"contig": {"contigName": "1", "contigLength": 249250621, "contigMD5": "1b22b98cdeb4a9304cb5d48026a85128", "referenceURL": "http:\/\/www.broadinstitute.org\/ftp\/pub\/seq\/references\/Homo_sapiens_assembly19.fasta", "assembly": "GRCh37", "species": "Homo Sapiens"}, "start": 248262648, "oldPosition": null, "end": 248262721, "mapq": 23, "readName": "C0FPVACXX120305:4:1303:12410:97560", "sequence": "GATCTTTTCAACAGTTACAGCAGAAAGTTTTCATGGAGAAATGGAATCACACTTCAAATGATTTCATTTTGTTGGG", "qual": "IBBHEFFEKFCKFHFACKFIJFJDCFHFEEDJBCHIFIDDBCGJDBBJAJBJFCIDCACHBDEBHADDDADDAED;", "cigar": "4S1M1D71M", "oldCigar": null, "basesTrimmedFromStart": 0, "basesTrimmedFromEnd": 0, "readPaired": true, "properPair": true, "readMapped": true, "mateMapped": true, "firstOfPair": false, "secondOfPair": true, "failedVendorQualityChecks": false, "duplicateRead": false, "readNegativeStrand": true, "mateNegativeStrand": false, "primaryAlignment": true, "secondaryAlignment": false, "supplementaryAlignment": false, "mismatchingPositions": "3^C71", "origQual": "JJJJIJJJJJJJJJIJJJJJJJJJJJJJJJJJJIJJJJJJJIJJJJIHFHHJJJJJJJJJJJJHHHHHFFFFFCCC", "attributes": "UQ:i:0\tOQ:Z:JJJJIJJJJJJJJJIJJJJJJJJJJJJJJJJJJIJJJJJJJIJJJJIHFHHJJJJJJJJJJJJHHHHHFFFFFCCC\tMQ:i:25\tSM:i:23\tNM:i:1\tAM:i:23\tRG:Z:C0FPV.4\tPG:Z:bwa.6\tTLEN:i:-70", "recordGroupName": "C0FPV.4", "recordGroupSequencingCenter": "BI", "recordGroupDescription": null, "recordGroupRunDateEpoch": 1330923600000, "recordGroupFlowOrder": null, "recordGroupKeySequence": null, "recordGroupLibrary": "Pond-140867", "recordGroupPredictedMedianInsertSize": null, "recordGroupPlatform": "illumina", "recordGroupPlatformUnit": "C0FPVACXX120305.4.TGTAACTC", "recordGroupSample": "JJN-3", "mateAlignmentStart": 248262650, "mateAlignmentEnd": null, "mateContig": {"contigName": "1", "contigLength": 249250621, "contigMD5": "1b22b98cdeb4a9304cb5d48026a85128", "referenceURL": "http:\/\/www.broadinstitute.org\/ftp\/pub\/seq\/references\/Homo_sapiens_assembly19.fasta", "assembly": "GRCh37", "species": "Homo Sapiens"}})
	at org.bdgenomics.adam.rich.DecadentRead$.apply(DecadentRead.scala:40)
	at org.bdgenomics.adam.rich.DecadentRead$.apply(DecadentRead.scala:32)
	at org.bdgenomics.adam.rich.DecadentRead$$anonfun$cloy$1.apply(DecadentRead.scala:50)
	at org.bdgenomics.adam.rich.DecadentRead$$anonfun$cloy$1.apply(DecadentRead.scala:50)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
	at scala.collection.Iterator$$anon$14.hasNext(Iterator.scala:389)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:144)
	at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.aggregate(TraversableOnce.scala:201)
	at scala.collection.AbstractIterator.aggregate(Iterator.scala:1157)
	at org.apache.spark.rdd.RDD$$anonfun$23.apply(RDD.scala:918)
	at org.apache.spark.rdd.RDD$$anonfun$23.apply(RDD.scala:918)
	at org.apache.spark.SparkContext$$anonfun$30.apply(SparkContext.scala:1389)
	at org.apache.spark.SparkContext$$anonfun$30.apply(SparkContext.scala:1389)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:61)
	at org.apache.spark.scheduler.Task.run(Task.scala:56)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:200)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: scala.MatchError: (D,Some(3)) (of class scala.Tuple2)
	at org.bdgenomics.adam.util.MdTag$.apply(MdTag.scala:70)
	at org.bdgenomics.adam.rich.RichAlignmentRecord.mdTag$lzycompute(RichAlignmentRecord.scala:98)
	at org.bdgenomics.adam.rich.RichAlignmentRecord.mdTag(RichAlignmentRecord.scala:96)
	at org.bdgenomics.adam.rich.RichAlignmentRecord.getReferenceBase$1(RichAlignmentRecord.scala:180)
	at org.bdgenomics.adam.rich.RichAlignmentRecord.getReferenceContext(RichAlignmentRecord.scala:196)
	at org.bdgenomics.adam.rich.RichAlignmentRecord$$anonfun$1$$anonfun$2.apply(RichAlignmentRecord.scala:217)
	at org.bdgenomics.adam.rich.RichAlignmentRecord$$anonfun$1$$anonfun$2.apply(RichAlignmentRecord.scala:215)

```

--- a/adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala
+++ b/adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala
@@ -17,6 +17,7 @@
  */
 package org.bdgenomics.adam.cli
 
+import htsjdk.samtools.ValidationStringency
 import org.apache.spark.rdd.RDD
 import org.apache.spark.{ Logging, SparkContext }
 import org.bdgenomics.adam.algorithms.consensus._
@@ -49,6 +50,8 @@ class TransformArgs extends Args4jBase with ADAMSaveAnyArgs with ParquetArgs {
   var markDuplicates: Boolean = false
   @Args4jOption(required = false, name = "-recalibrate_base_qualities", usage = "Recalibrate the base quality scores (ILLUMINA only)")
   var recalibrateBaseQualities: Boolean = false
+  @Args4jOption(required = false, name = "-strict_bqsr", usage = "Run BQSR with strict validation.")
+  var strictBQSR: Boolean = false
   @Args4jOption(required = false, name = "-dump_observations", usage = "Local path to dump BQSR observations to. Outputs CSV format.")
   var observationsPath: String = null
   @Args4jOption(required = false, name = "-known_snps", usage = "Sites-only VCF giving location of known SNPs")
@@ -118,7 +121,14 @@ class Transform(protected val args: TransformArgs) extends BDGSparkCommand[Trans
     if (args.recalibrateBaseQualities) {
       log.info("Recalibrating base qualities")
       val knownSnps: SnpTable = createKnownSnpsTable(sc)
-      adamRecords = adamRecords.adamBQSR(sc.broadcast(knownSnps), Option(args.observationsPath))
+      val stringency = if (args.strictBQSR) {
+        ValidationStringency.STRICT
+      } else {
+        ValidationStringency.LENIENT
+      }
+      adamRecords = adamRecords.adamBQSR(sc.broadcast(knownSnps),
+        Option(args.observationsPath),
+        stringency)
     }
 
     if (args.coalesce != -1) {

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala
@@ -299,8 +299,9 @@ class AlignmentRecordRDDFunctions(rdd: RDD[AlignmentRecord])
    * @return Returns an RDD of recalibrated reads.
    */
   def adamBQSR(knownSnps: Broadcast[SnpTable],
-               observationDumpFile: Option[String] = None): RDD[AlignmentRecord] = BQSRInDriver.time {
-    BaseQualityRecalibration(rdd, knownSnps, observationDumpFile)
+               observationDumpFile: Option[String] = None,
+               validationStringency: ValidationStringency = ValidationStringency.LENIENT): RDD[AlignmentRecord] = BQSRInDriver.time {
+    BaseQualityRecalibration(rdd, knownSnps, observationDumpFile, validationStringency)
   }
 
   /**

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/BaseQualityRecalibration.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/BaseQualityRecalibration.scala
@@ -17,6 +17,7 @@
  */
 package org.bdgenomics.adam.rdd.read.recalibration
 
+import htsjdk.samtools.ValidationStringency
 import java.io._
 import org.apache.spark.SparkContext._
 import org.apache.spark.Logging
@@ -35,7 +36,7 @@ import org.bdgenomics.formats.avro.AlignmentRecord
  * quality scores.
  */
 class BaseQualityRecalibration(
-  val input: RDD[DecadentRead],
+  val input: RDD[(Option[DecadentRead], Option[AlignmentRecord])],
   val knownSnps: Broadcast[SnpTable],
   val dumpObservationTableFile: Option[String] = None)
     extends Serializable with Logging {
@@ -68,7 +69,7 @@ class BaseQualityRecalibration(
       covariates(read).zip(read.residues).
         filter { case (key, residue) => shouldIncludeResidue(residue) }
 
-    input.filter(shouldIncludeRead).flatMap(observe)
+    input.flatMap(_._1).filter(shouldIncludeRead).flatMap(observe)
   }
 
   if (enableVisitLogging) {
@@ -92,7 +93,7 @@ class BaseQualityRecalibration(
 
   val result: RDD[AlignmentRecord] = {
     val recalibrator = Recalibrator(observed, minAcceptableQuality)
-    input.map(recalibrator)
+    input.map(recalibrator(_))
   }
 
   private def dumpVisits(filename: String) = {
@@ -103,7 +104,7 @@ class BaseQualityRecalibration(
         (if (read.record.getSecondOfPair) "2" else "")
 
     val readLengths =
-      input.map(read => (readId(read), read.residues.length)).collectAsMap()
+      input.flatMap(_._1).map(read => (readId(read), read.residues.length)).collectAsMap()
 
     val visited = dataset.
       map { case (key, residue) => (readId(residue.read), Seq(residue.offset)) }.
@@ -125,6 +126,7 @@ class BaseQualityRecalibration(
 object BaseQualityRecalibration {
   def apply(rdd: RDD[AlignmentRecord],
             knownSnps: Broadcast[SnpTable],
-            observationDumpFile: Option[String] = None): RDD[AlignmentRecord] =
-    new BaseQualityRecalibration(cloy(rdd), knownSnps).result
+            observationDumpFile: Option[String] = None,
+            validationStringency: ValidationStringency = ValidationStringency.STRICT): RDD[AlignmentRecord] =
+    new BaseQualityRecalibration(cloy(rdd, validationStringency), knownSnps).result
 }

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/Recalibrator.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/Recalibrator.scala
@@ -26,18 +26,20 @@ import org.bdgenomics.adam.instrumentation.Timers._
 import scala.math.{ exp, log }
 
 class Recalibrator(val table: RecalibrationTable, val minAcceptableQuality: QualityScore)
-    extends (DecadentRead => AlignmentRecord) with Serializable {
+    extends Serializable {
 
-  def apply(read: DecadentRead): AlignmentRecord = RecalibrateRead.time {
-    val record: AlignmentRecord = read.record
-    if (record.getQual != null) {
-      AlignmentRecord.newBuilder(record)
-        .setQual(QualityScore.toString(computeQual(read)))
-        .setOrigQual(record.getQual)
-        .build()
-    } else {
-      record
-    }
+  def apply(r: (Option[DecadentRead], Option[AlignmentRecord])): AlignmentRecord = RecalibrateRead.time {
+    r._1.fold(r._2.get)(read => {
+      val record: AlignmentRecord = read.record
+      if (record.getQual != null) {
+        AlignmentRecord.newBuilder(record)
+          .setQual(QualityScore.toString(computeQual(read)))
+          .setOrigQual(record.getQual)
+          .build()
+      } else {
+        record
+      }
+    })
   }
 
   def computeQual(read: DecadentRead): Seq[QualityScore] = ComputeQualityScore.time {

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rich/DecadentRead.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rich/DecadentRead.scala
@@ -17,6 +17,7 @@
  */
 package org.bdgenomics.adam.rich
 
+import htsjdk.samtools.ValidationStringency
 import org.apache.spark.Logging
 import org.apache.spark.rdd.RDD
 import org.bdgenomics.adam.models.ReferencePosition
@@ -25,7 +26,7 @@ import org.bdgenomics.adam.rich.RichAlignmentRecord._
 import org.bdgenomics.adam.util.{ MdTag, QualityScore }
 import org.bdgenomics.formats.avro.AlignmentRecord
 
-object DecadentRead {
+object DecadentRead extends Logging with Serializable {
   type Residue = DecadentRead#Residue
 
   // Constructors
@@ -47,7 +48,25 @@ object DecadentRead {
    *   2. To clog, to glut, or satisfy, as the appetite; to satiate.
    *   3. To fill up or choke up; to stop up.
    */
-  def cloy(rdd: RDD[AlignmentRecord]): RDD[DecadentRead] = rdd.map(DecadentRead.apply)
+  def cloy(rdd: RDD[AlignmentRecord],
+           strictness: ValidationStringency = ValidationStringency.STRICT): RDD[(Option[DecadentRead], Option[AlignmentRecord])] = {
+    rdd.map(r => {
+      try {
+        val dr = DecadentRead.apply(r)
+        (Some(dr), None)
+      } catch {
+        case e: Throwable => {
+          if (strictness == ValidationStringency.STRICT) {
+            throw e
+          } else {
+            log.warn("Converting read %s to decadent read failed with %s. Skipping...".format(
+              r, e))
+            (None, Some(r))
+          }
+        }
+      }
+    })
+  }
 
   // The inevitable counterpart of the above.
   implicit def decay(rdd: RDD[DecadentRead]): RDD[AlignmentRecord] = rdd.map(_.record)

--- a/adam-core/src/test/scala/org/bdgenomics/adam/rich/DecadentReadSuite.scala
+++ b/adam-core/src/test/scala/org/bdgenomics/adam/rich/DecadentReadSuite.scala
@@ -18,11 +18,14 @@
 
 package org.bdgenomics.adam.rich
 
-import org.scalatest.FunSuite
+import htsjdk.samtools.ValidationStringency
+import org.apache.spark.SparkException
+import org.apache.spark.rdd.RDD
 import org.bdgenomics.adam.models.ReferencePosition
+import org.bdgenomics.adam.util.ADAMFunSuite
 import org.bdgenomics.formats.avro.{ AlignmentRecord, Contig }
 
-class DecadentReadSuite extends FunSuite {
+class DecadentReadSuite extends ADAMFunSuite {
 
   test("reference position of decadent read") {
     val contig = Contig.newBuilder
@@ -84,5 +87,75 @@ class DecadentReadSuite extends FunSuite {
 
     val record = DecadentRead(hardClippedRead)
     assert(record.residues.size === 9)
+  }
+
+  test("converting bad read should fail") {
+    val readBad = AlignmentRecord.newBuilder()
+      .setContig(Contig.newBuilder()
+        .setContigName("1")
+        .build())
+      .setStart(248262648L)
+      .setEnd(248262721L)
+      .setMapq(23)
+      .setSequence("GATCTTTTCAACAGTTACAGCAGAAAGTTTTCATGGAGAAATGGAATCACACTTCAAATGATTTCATTTTGTTGGG")
+      .setQual("IBBHEFFEKFCKFHFACKFIJFJDCFHFEEDJBCHIFIDDBCGJDBBJAJBJFCIDCACHBDEBHADDDADDAED;")
+      .setCigar("4S1M1D71M")
+      .setReadMapped(true)
+      .setMismatchingPositions("3^C71")
+      .build()
+
+    intercept[IllegalArgumentException] {
+      DecadentRead(readBad)
+    }
+  }
+
+  def badGoodReadRDD: RDD[AlignmentRecord] = {
+    val readBad = AlignmentRecord.newBuilder()
+      .setContig(Contig.newBuilder()
+        .setContigName("1")
+        .build())
+      .setStart(248262648L)
+      .setEnd(248262721L)
+      .setMapq(23)
+      .setSequence("GATCTTTTCAACAGTTACAGCAGAAAGTTTTCATGGAGAAATGGAATCACACTTCAAATGATTTCATTTTGTTGGG")
+      .setQual("IBBHEFFEKFCKFHFACKFIJFJDCFHFEEDJBCHIFIDDBCGJDBBJAJBJFCIDCACHBDEBHADDDADDAED;")
+      .setCigar("4S1M1D71M")
+      .setReadMapped(true)
+      .setMismatchingPositions("3^C71")
+      .build()
+    val readGood = AlignmentRecord.newBuilder()
+      .setContig(Contig.newBuilder()
+        .setContigName("1")
+        .build())
+      .setStart(248262648L)
+      .setEnd(248262721L)
+      .setMapq(23)
+      .setSequence("GATCTTTTCAACAGTTACAGCAGAAAGTTTTCATGGAGAAATGGAATCACACTTCAAATGATTTCATTTTGTTGGG")
+      .setQual("IBBHEFFEKFCKFHFACKFIJFJDCFHFEEDJBCHIFIDDBCGJDBBJAJBJFCIDCACHBDEBHADDDADDAED;")
+      .setCigar("4S1M1D71M")
+      .setReadMapped(true)
+      .setMismatchingPositions("1^C71")
+      .build()
+    sc.parallelize(Seq(readBad, readGood))
+  }
+
+  sparkTest("convert an RDD that has an bad read in it with loose validation") {
+    val rdd = badGoodReadRDD
+    val decadent = DecadentRead.cloy(rdd, ValidationStringency.LENIENT)
+      .repartition(2)
+      .collect()
+    assert(decadent.size === 2)
+    assert(decadent.count(_._1.isDefined) === 1)
+    assert(decadent.count(_._1.isEmpty) === 1)
+    assert(decadent.count(_._2.isDefined) === 1)
+    assert(decadent.count(_._2.isEmpty) === 1)
+  }
+
+  sparkTest("converting an RDD that has an bad read in it with strict validation will throw an error") {
+    val rdd = badGoodReadRDD
+    intercept[SparkException] {
+      // need count to force computation
+      DecadentRead.cloy(rdd).count
+    }
   }
 }

--- a/adam-core/src/test/scala/org/bdgenomics/adam/util/MdTagSuite.scala
+++ b/adam-core/src/test/scala/org/bdgenomics/adam/util/MdTagSuite.scala
@@ -602,4 +602,10 @@ class MdTagSuite extends FunSuite {
   test("handle '=' and 'X' operators") {
     testTag("ACCCAAGT", "ACCATAGA", "3=2X2=1X", 0, "3A0T2A0", 0, 7)
   }
+
+  test("CIGAR/MD tag mismatch should cause errors") {
+    intercept[MatchError] {
+      MdTag("3^C71", 1L, TextCigarCodec.decode("4S1M1D71M"))
+    }
+  }
 }

